### PR TITLE
Gradient continuity InterfaceKernel 

### LINKED ISSUE #8211: Interface kernel to enforce gradient continuity across internal boundaries
### Description of the enhancement or error report
To enforce a diffusion flux across a boundary (or a gradient continuity) a new InterfaceKernel is required.

### Rationale for the enhancement or information for reproducing the error
This will be a useful example for the application if InterfaceKernels, and a component in a new strain periodicity system based on InterfaceKernels.

### Identified impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Add InterfaceKernel to enforce gradient continuity.

Closes #8211

### CODE REVIEW

File: modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      return -_grad_phi_neighbor[_j][_qp](_component) * _test_neighbor[_i][_qp];
+
+    case Moose::ElementNeighbor:
+      return -_grad_phi_neighbor[_j][_qp](_component) * _test[_i][_qp];
+
+    case Moose::NeighborElement:
+      return _grad_phi[_j][_qp](_component) * _test_neighbor[_i][_qp];
+  }
+
+  mooseError("Internal error.");
```
Comment (fdkong) [inline modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C]: Could we add something meaningful like: "do not support DGJacobianType %s"?

**Review by fdkong — CHANGES_REQUESTED** [review]

File: modules/phase_field/tests/misc/interface_flux.i [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+#    variable = u1
+#    boundary = 10
+#  [../]
+#  [./v_boundary_term]
+#    type = DiffusionFluxBC
+#    variable = v1
+#    boundary = 11
+#  [../]
+#[]
+
```
Comment (fdkong) [inline modules/phase_field/tests/misc/interface_flux.i]: Remove comments?

**Review by fdkong — CHANGES_REQUESTED** [review]

File: modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C]: There is no unsupported option. The error is only needed to satisfy gcc. Clang is smart enough to realize that line is unreachable and the function always returns. And I could not convert that enum into a string anyways.

File: modules/phase_field/tests/misc/interface_flux.i (hunk 2 above)
Comment (dschwen) [inline modules/phase_field/tests/misc/interface_flux.i]: See top comment in that file. I'd like to retain the objects that reproduce the same result. But I didn't want to bloat the gold file :-/

File: modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    _component(getParam<unsigned int>("component"))
+{
+}
+
+Real
+InterfaceGradientMatch::computeQpResidual(Moose::DGResidualType type)
+{
+  // equal gradients means difference is zero
+  Real res = _grad_u[_qp](_component) - _grad_neighbor_value[_qp](_component);
+
```
Comment (fdkong) [inline modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C]: More general way is to subtract along the normal direction of the surface, I guess. Do you assume that the boundary will be aligned  with one of axis directions?

File: modules/phase_field/examples/interfacekernels/interface_gradient.i [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    boundary = 10
+    neighbor_var = v
+    component = 0
+  [../]
+  [./diffusion_surface_term]
+    type = InterfaceDiffusionFlux
+    boundary = 10
+    variable = u
+    neighbor_var = v
+  [../]
```
Comment (fdkong) [inline modules/phase_field/examples/interfacekernels/interface_gradient.i]: How these two boundary conditions coexist? Will we get a totally different solution if we remove the second condition  `diffusion_surface_term`.
 I am  curious.

File: modules/phase_field/tests/misc/interface_grad.i [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+[InterfaceKernels]
+  [./iface]
+    type = InterfaceGradientMatch
+    variable = u
+    boundary = 10
+    neighbor_var = v
+    component = 0
+  [../]
+[]
+
```
Comment (fdkong) [inline modules/phase_field/tests/misc/interface_grad.i]: This makes sense.

File: modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C (hunk 3 above)
Comment (dschwen) [inline modules/phase_field/src/interfacekernels/InterfaceGradientMatch.C]: I assume that I have to apply _n_ kernels for each of the gradient components (to get the periodicity of the entire strain tensor for example).

File: modules/phase_field/src/interfacekernels/InterfaceDiffusionFlux.C [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+Real
+InterfaceDiffusionFlux::computeQpResidual(Moose::DGResidualType type)
+{
+  // add weak form surface terms for the diffusion equation
+  switch (type)
+  {
+    case Moose::Element:
+      return -_D * _grad_u[_qp] * _normals[_qp] * _test[_i][_qp];
```
Comment (dschwen) [inline modules/phase_field/src/interfacekernels/InterfaceDiffusionFlux.C]: I would like to keep this as a simple example that demonstrates the relation between integrated BCs and interface kernels.

Comment (YaqiWang) [inline modules/phase_field/src/interfacekernels/InterfaceDiffusionFlux.C]: Interface kernels should be closely related with DG kernels, except they allow the variables change on the two sides and do not necessarily operate on all internal sides.

Comment (dschwen) [inline modules/phase_field/src/interfacekernels/InterfaceDiffusionFlux.C]: Where are you going with that comment?

**Review by fdkong — APPROVED** [review]

**Pull request merged**